Sync api and support moving gateways away from aci

sebageek opened the pull request:
Sync networks, move l3 gateways. Basically a prerequisite for https://github.com/sapcc/networking-ccloud/pull/112

sebageek committed:
Use request.context in API where possible

Each request already provides a context, no need that we create our own
each time.

sebageek committed:
Add API to sync a network

The ACI API now allows a user to sync a network by doing a PUT on the
respective resource. This will internally call sync_network(), which in
the end will end up in the same call the ACI syncloop uses.

sebageek committed:
Allow gateways to move off ACI

The ACI driver now has a config switch that indicates if it is
responsible for all l3 gateways. If turned off, we will not configure
any l3 gateways on ACI.

The other option for a network to not have its gateway on ACI is by it
being tagged with "gateway-host::cc-fabric". networking-ccloud will
throw an event whenever it wants a gateway to be moved from ACI to the
cc-fabric. We react on that event by tearing down all subnets from that
network. The network will still have some different properties set, as
it is still handled as an external network (but without any l3
gateways). We'll see if that is a good idea or if this is something
that'll need to change. Experience will show.

joker-at-work commented on networking_aci/plugins/ml2/drivers/mech_aci/driver.py:
```diff
@@ -265,6 +268,10 @@ def create_subnet_postcommit(self, context):
 
     def delete_subnet_postcommit(self, context):
+        if not CONF.ml2_aci.handle_all_l3_gateways or \
+                aci_const.CC_FABRIC_L3_GATEWAY_TAG in context.network.get('tags', []):
```
why is this a `.get('tags', [])`, but the others access it as if it's always there i.e. `network['tags']`

sebageek replied:
Yeah, well, I guess because the original aci driver was written with a lot of gets and I just... let myself be persuaded by that. Shouldn't be a get here.

joker-at-work commented on networking_aci/plugins/ml2/drivers/mech_aci/driver.py:
```diff
@@ -300,6 +307,37 @@ def on_fabric_transit_created(self, resource, event, trigger, payload):
+        for n, subnet_id in enumerate(network['subnets']):
+            subnet = self.get_subnet(payload.context, subnet_id)
+            subnetpool_id = subnet['subnetpool_id']
+            address_scope_name = self.db.get_address_scope_name(context._plugin_context, subnetpool_id)
+            if address_scope_name is None:
+                continue
+            last_on_network = n + 1 == len(network['subnets'])
+            self.rpc_notifier.delete_subnet(subnet, external=True, address_scope_name=address_scope_name,
+                                            last_on_network=last_on_network)
```
Is it important that we might not have sent out a delete for subnets without address_scope_name and thus `last_on_network` is maybe not actually the last on the network, just the last in the list of subnets?

sebageek replied:
That's actually a good point. I'm not 100% sure on the whole "last_network" semantic, I think in the end it'll delete the l3out and stuff and I'm not sure that this is something the syncloop will properly do.... So, I think I'll do an extra iteration over it. :)

sebageek committed:
Pass core plugin, not l3 plugin to API controllers

All controllers got the l3 plugin as an argument. As nobody needed the
l3 plugin and the NetwworksController needed the core plugin, I removed
the plugin as an argument from the other plugins and gave
NetwworksController the right plugin.
